segmentation_models_pytorch/decoders/deeplabv3/decoder.py

```diff
@@ -106,17 +100,10 @@ def __init__(
             nn.ReLU(),
         )
 
-        scale_factor = 2 if output_stride == 8 else 4
+        scale_factor = 4 if output_stride == 16 and encoder_depth > 3 else 2
         self.up = nn.UpsamplingBilinear2d(scale_factor=scale_factor)
 
-        if encoder_depth == 3 and output_stride == 8:
-            self.highres_input_index = -2
-        elif encoder_depth == 3 or encoder_depth == 4:
-            self.highres_input_index = -3
-        else:
-            self.highres_input_index = -4
-
-        highres_in_channels = encoder_channels[self.highres_input_index]
+        highres_in_channels = encoder_channels[2]
```

Comment:
The only concern here is speed, cause the higher resolution feature we take, the more we need to process

Reply:
I think there might have been some misunderstanding. 

Currently, regardless of the `encoder_depth`, `high_res_features` is always fixed at 1/4 of the input resolution, which improves speed.

Before #986 and #991, the resolution of `high_res_features` varied depending on the `encoder_depth`:
- `encoder_depth=5`: `high_res_features` was 1/4 of the input resolution.
- `encoder_depth=4`: `high_res_features` was 1/2 of the input resolution.
- `encoder_depth=3`: `high_res_features` matched the input resolution.

Reply:
For your reference, in PR #986, I updated the `high_res_features` logic as follows:

- When `encoder_depth`=5: `encoder_channels[-4]`, which corresponds to index 2 (1/4 of the input resolution).
- When `encoder_depth`=4: `encoder_channels[-3]`, which corresponds to index 2 (1/4 of the input resolution).
- When `encoder_depth`=3:
  - If `encoder_output_stride`=8: `encoder_channels[-2]`, which corresponds to index 3 (1/2 of the input resolution).
  - If `encoder_output_stride`=16: `encoder_channels[-3]`, which corresponds to index 2 (1/4 of the input resolution).

The only difference in `high_res_features` is when `encoder_depth`=3 and `encoder_output_stride`=8.
I think PR #991 is preferable because it allows you to consistently set `upsampling` to 4 to preserve the input/output tensor sizes.  In contrast, PR #986 requires setting `upsampling` to 2 to maintain sizes when `encoder_depth`=3 and `encoder_output_stride`=8.

Reply:
Thanks for explanations! It's clearer now 🙏